pkg/components/renderer/renderer.go

```diff
@@ -1,159 +1,256 @@
+}
+
+func getLocalDomain() string {
 	if setting.HttpAddr != setting.DEFAULT_HTTP_ADDR {
-		localDomain = setting.HttpAddr
+		return setting.HttpAddr
 	}
 
-	url := fmt.Sprintf("%s://%s:%s/%s", setting.Protocol, localDomain, setting.HttpPort, params.Path)
+	return "localhost"
+}
 
-	binPath, _ := filepath.Abs(filepath.Join(setting.PhantomDir, executable))
-	scriptPath, _ := filepath.Abs(filepath.Join(setting.PhantomDir, "render.js"))
-	pngPath, _ := filepath.Abs(filepath.Join(setting.ImagesDir, util.GetRandomString(20)))
-	pngPath = pngPath + ".png"
+func getURL(path string) string {
+	return fmt.Sprintf("%s://%s:%s/%s", setting.Protocol, getLocalDomain(), setting.HttpPort, path)
+}
 
-	orgRole := params.OrgRole
-	if params.IsAlertContext {
-		orgRole = models.ROLE_ADMIN
-	}
-	renderKey := middleware.AddRenderAuthKey(params.OrgId, params.UserId, orgRole)
-	defer middleware.RemoveRenderAuthKey(renderKey)
+func getRenderKey(OrgID, UserID int64, OrgRole models.RoleType) string {
+	rendererLog.Debug("adding render authkey", "orgid", OrgID, "userid", UserID, "role", OrgRole)
+	return middleware.AddRenderAuthKey(OrgID, UserID, OrgRole)
+}
 
-	timeout, err := strconv.Atoi(params.Timeout)
-	if err != nil {
-		timeout = 15
+func getExecPath() (string, error) {
+	if setting.RendererChromiumExecPath != "" {
+		return setting.RendererChromiumExecPath, nil
 	}
 
-	phantomDebugArg := "--debug=false"
-	if log.GetLogLevelFor("png-renderer") >= log.LvlDebug {
-		phantomDebugArg = "--debug=true"
+	if setting.RendererChromiumRevision == "" {
+		return "", fmt.Errorf("Revision of bundled chromium unknown you need to manually set exec path")
 	}
 
-	cmdArgs := []string{
-		"--ignore-ssl-errors=true",
-		"--web-security=false",
-		phantomDebugArg,
-		scriptPath,
-		"url=" + url,
-		"width=" + params.Width,
-		"height=" + params.Height,
-		"png=" + pngPath,
-		"domain=" + localDomain,
-		"timeout=" + strconv.Itoa(timeout),
-		"renderKey=" + renderKey,
+	basePath := filepath.Join(setting.HomePath, "tools/chromium", setting.RendererChromiumRevision)
+	if runtime.GOOS == "windows" {
+		// the same for both 32 and 64bit
+		return filepath.Join(basePath, "chrome-win32/chrome.exe"), nil
+	} else if runtime.GOOS == "darwin" {
+		return filepath.Join(basePath, "chrome-mac/Chromium.app/Contents/MacOS/Chromium"), nil
+	} else if runtime.GOOS == "linux" {
+		return filepath.Join(basePath, "chrome-linux/chrome"), nil
 	}
 
-	if params.Encoding != "" {
-		cmdArgs = append([]string{fmt.Sprintf("--output-encoding=%s", params.Encoding)}, cmdArgs...)
+	return "", fmt.Errorf("Unexpected OS %s you need to manually set chromium exec path", runtime.GOOS)
+}
+
+func screenshotAction(width, height int, url, renderKey, pngPath string) chromedp.Tasks {
+	var buf []byte
+	return chromedp.Tasks{
+		setViewportAction(width, height),
+		setCookieAction(getLocalDomain(), "renderKey", renderKey),
+		chromedp.Navigate(url),
+		chromedp.WaitReady(`.renderingComplete`, chromedp.ByQuery),
```

Comment:
This will work for single panel rendering only. Please note that currently phantomjs can also do full dashboard rendering /render/full/path/to/dashboard.
Ideally we'd like to keep this support available.

Reply:
Ok will fix.

In order for this to work I will add a `.dashboardRenderingComplete` can you point me a place in the code where such an event would be added?